salesagility/SuiteCRM
# Add basic Seeds and Factories system

## Description

This is a minimum viable implementation of the seeding and factories systems I described in [this forum post](https://community.suitecrm.com/t/database-seeding/70823/3) a few days ago.

This is still a work in progress, but I'm opening this to get feedback and see if people like the idea.

An example of how this can be used:
```php
<?php

# The factories can be used in a few ways.

# With no arguments, the `create` method will create a User with default properties.
$factory = new UserFactory();
$userBean = $factory->create(); # This returns an array with one UserBean.

echo $userBean[0]->first_name; #=> 'jane', or some other name generated by Faker

# With an argument that passes an array of properties, the `create` method
# will create a User with properties based on the passed array, and default
# properties for any others that aren't explicitly included in this
# argument.
$factory = new UserFactory();
$userBean = $factory->create([
  'user_name' => 'foobar',
  'first_name' => 'foo',
  'last_name' => 'bar'
]);

echo $userBean[0]->first_name; #=> 'foo'
```

If the factory doesn't include a property you need by default, you can just pass it as part of the `define` method. For example, to create a randomly-generated user that's an admin, you could do `$factory->create(['is_admin' => 1]);`.

Right now, the `db:seed` Robo task will create 1 user with default properties (randomly generated by Faker).

TODO:
- [x] Rename `define` to something else?
- [ ] Figure out a better place to put this code, I'd rather not have a top-level `seeds/` directory.
- [ ] Add tests for the factories.
- [ ] Open a PR in the SuiteDocs repo documenting how this works/how to use it.
- [x] Add something to the factories so they can be used to create more than one instance of an item at a time.
- [ ] Update the `db:seed` Robo task so it accepts a `--only` flag, e.g. `--only=Accounts,Users`.
- [ ] Update a few tests to use the factories, as a proof-of-concept?
- [ ] Possibly add a `getInstance()` method on BaseFactory so it can be used like `UserFactory::getInstance()->create()`.
- [x] Probably add at least other factory, maybe for Accounts?
- [ ] Open an issue for tracking various improvements to this system, since I have quite a few things I'd like to improve about it that aren't really necessary for the initial implementation.
- [ ] Possibly add validation to this, to error if people are trying to pass invalid parameters to the given factory (e.g. `user_nam` or `username` instead of `user_name`).
- [ ] States like in Laravel's factories, though those aren't necessary for this PR.

## Motivation and Context

As I said in the forum post where I pitched this, the system is based on Laravel's [Factories](https://laravel.com/docs/6.x/database-testing#writing-factories) and [Seeding](https://laravel.com/docs/6.x/seeding) systems. I borrowed some of the code from Laravel when building this.

The main use cases for these factories are for seeding and for use in automated testing, to create randomly-generated beans to work from.

## How To Test This

Run `./vendor/bin/robo db:seed` and see if it creates a new User in the database.

## Types of changes

### Final checklist
